pkg/receive/writer.go

```diff
@@ -43,7 +44,16 @@ func (r *Writer) Receive(wreq *prompb.WriteRequest) error {
 
 		for _, s := range t.Samples {
 			_, err = app.Add(lset, s.Timestamp, s.Value)
+			switch err {
+			case storage.ErrOutOfOrderSample:
+				level.Debug(r.logger).Log("msg", "Out of order sample", "lset", lset.String(), "sample", s.String())
+			case storage.ErrDuplicateSampleForTimestamp:
+				level.Debug(r.logger).Log("msg", "Duplicate sample for timestamp", "lset", lset.String(), "sample", s.String())
+			case storage.ErrOutOfBounds:
+				level.Debug(r.logger).Log("msg", "Out of bounds metric", "lset", lset.String(), "sample", s.String())
+			}
 			if err != nil {
+				app.Rollback()
```

Comment:
why do you need to rollback here?

Imagine the situation:
added sample 1
error on adding sample2

when you run a rollback you also discard sample 1

when you hit an error at app.Add it doesn't get added to the head anyway. I don't see why you need to run a rollback. Correct me if my assumption is wrong.

Reply:
If my understanding of the mechanism is correct, then on any failed sample send we should be discarding all samples appended in that batch. 

You are correct -- currently sample1 is added, but sample2 isn't. After returning a [500 error code](https://github.com/prometheus/prometheus/blob/08c55c119f39093e18b2bb9cba5c5619dc4ea0e1/storage/remote/client.go#L112-L114), Prometheus considers the batch of samples sent recoverable, and retries all samples -- see [sendSamplesWithBackoff](https://github.com/prometheus/prometheus/blob/3344bb5c332838df21f87b6473c0764315f5ca71/storage/remote/queue_manager.go#L781). On the next retry there is now a conflict on sample1 since you already appended it.

Reply:
From my POV we only want to notify about this conflict and not cause Prometheus to try and resend these as sending the same samples will result in the same conflict. 
As mentioned in the original issue the remote write shouldn't even send these and it this should be fixed in Prometheus as well.

The idea with this change is to  ingest the samples and just ignore the conflict errors as these are harmless and notify Prometheus with `http.StatusConflict`. 
Now I haven't looked at the  Prometheus code for the remote write, but if it tries to resend the same samples even with the `http.StatusConflict` than a PR should be opened to change this behavior in Prometheus.

https://github.com/thanos-io/thanos/issues/1509#issuecomment-531261715
>What error code does Thanos return when failing due to an out of order error? Prometheus should only retry on 5XX errors, so returning a 4XX would probably be the correct behavior and avoid the retries.

@csmarchbanks's comment suggests that `4XX` are not retried so this change should work as expected.

I haven't looked at the tests, but I would definitely prefer to have some tests to ensure this behavior.

Reply:
👍 Yes, receiving a 4xx code results in Prometheus abandoning the batch and moving onto the next.

I updated my commit to reflect that we should add all samples possible even when encountering errors.